File: masari-platform/translate.mjs
import fs from 'fs';
import { translate } from '@vitalets/google-translate-api';

async function chunkList(lst, n) {
  const chunks = [];
  for (let i = 0; i < lst.length; i += n) {
    chunks.push(lst.slice(i, i + n));
  }
  return chunks;
}

async function main() {
  const filePath = 'src/data/jobs300.js';
  let content = fs.readFileSync(filePath, 'utf8');

  // Skip if already translated
  if (content.includes('arTitle:')) {
    console.log('Already translated!');
    return;
  }

  const titleRegex = /title:\s*'(.*?)'/g;
  const descRegex = /description:\s*'(.*?)'/g;

  const titles = [];
  const descs = [];
  let match;

  while ((match = titleRegex.exec(content)) !== null) {
    titles.push(match[1]);
  }
  while ((match = descRegex.exec(content)) !== null) {
    descs.push(match[1]);
  }

  if (titles.length === 0) {
    console.log('No titles found.');
    return;
  }

  console.log(`Translating ${titles.length} titles and ${descs.length} descriptions...`);
  
  const arTitles = [];
  const arDescs = [];

  const titleChunks = await chunkList(titles, 30);
  for (const chunk of titleChunks) {
    const text = chunk.join('\n ||| \n');
    try {
      const res = await translate(text, { to: 'ar' });
      const parts = res.text.split('|||').map(t => t.trim());
      if (parts.length === chunk.length) {
        arTitles.push(...parts);
      } else { // fallback
        for (const item of chunk) {
          const single = await translate(item, { to: 'ar' });
          arTitles.push(single.text.trim());
        }
      }
    } catch (e) {
      console.log('Error translating titles block', e);
      for (const item of chunk) {
        const single = await translate(item, { to: 'ar' });
        arTitles.push(single.text.trim());
      }
    }
  }

  const descChunks = await chunkList(descs, 30);
  for (const chunk of descChunks) {
    const text = chunk.join('\n ||| \n');
    try {
      const res = await translate(text, { to: 'ar' });
      const parts = res.text.split('|||').map(t => t.trim());
      if (parts.length === chunk.length) {
        arDescs.push(...parts);
      } else {
        for (const item of chunk) {
          const single = await translate(item, { to: 'ar' });
          arDescs.push(single.text.trim());
        }
      }
    } catch (e) {
      console.log('Error translating descs block', e);
      for (const item of chunk) {
        const single = await translate(item, { to: 'ar' });
        arDescs.push(single.text.trim());
      }
    }
  }

  if (titles.length !== arTitles.length || descs.length !== arDescs.length) {
    console.log('Mismatch in lengths!', titles.length, arTitles.length, descs.length, arDescs.length);
    return;
  }

  let tCount = 0;
  content = content.replace(/title:\s*'(.*?)'/g, (match, p1) => {
    const t = arTitles[tCount].replace(/'/g, "\\'");
    tCount++;
    return `title: '${p1}', arTitle: '${t}'`;
  });

  let dCount = 0;
  content = content.replace(/description:\s*'(.*?)'/g, (match, p1) => {
    const d = arDescs[dCount].replace(/'/g, "\\'");
    dCount++;
    return `description: '${p1}', arDescription: '${d}'`;
  });

  fs.writeFileSync(filePath, content, 'utf8');
  console.log('Success: translations added');
}

main().catch(console.error);

File: masari-platform/translate.py
import re
from googletrans import Translator
import time
import os

def chunk_list(lst, n):
    for i in range(0, len(lst), n):
        yield lst[i:i + n]

def main():
    file_path = 'src/data/jobs300.js'
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Find all titles and descriptions
    titles = re.findall(r"title:\s*'(.*?)'", content)
    descs = re.findall(r"description:\s*'(.*?)'", content)
    
    if len(titles) == 0:
        print("No titles found.")
        return

    print(f"Translating {len(titles)} titles and {len(descs)} descriptions...")
    translator = Translator()
    
    ar_titles = []
    ar_descs = []
    
    try:
        # bulk translate titles
        for chunk in chunk_list(titles, 30):
            text = '\n<SEP>\n'.join(chunk)
            result = translator.translate(text, src='en', dest='ar')
            # It might translate <SEP> to Arabic characters if we aren't careful. 
            parts = result.text.split('<SEP>')
            if len(parts) == 1:
                # Fallback to translate items one by one 
                for item in chunk:
                    res = translator.translate(item, src='en', dest='ar')
                    ar_titles.append(res.text.strip())
            else:
                ar_titles.extend([t.strip() for t in parts])
            time.sleep(0.5)
            
        # bulk translate descriptions
        for chunk in chunk_list(descs, 30):
            text = '\n<SEP>\n'.join(chunk)
            result = translator.translate(text, src='en', dest='ar')
            parts = result.text.split('<SEP>')
            if len(parts) == 1:
                # Fallback
                for item in chunk:
                    res = translator.translate(item, src='en', dest='ar')
                    ar_descs.append(res.text.strip())
            else:
                ar_descs.extend([t.strip() for t in parts])
            time.sleep(0.5)
            
    except Exception as e:
        print(f"Translation failed: {e}")
        return

    # To ensure safety on splitting mismatch due to translating <SEP>
    if len(titles) != len(ar_titles) or len(descs) != len(ar_descs):
        print(f"Mismatch in translations length: {len(titles)} vs {len(ar_titles)}")
        print("Using item-by-item translation instead...")
        ar_titles = []
        ar_descs = []
        for t in titles:
            ar_titles.append(translator.translate(t, src='en', dest='ar').text)
            time.sleep(0.1)
        for d in descs:
            ar_descs.append(translator.translate(d, src='en', dest='ar').text)
            time.sleep(0.1)

    # Use iterator for replacement to handle duplicate items properly
    title_iter = iter(ar_titles)
    desc_iter = iter(ar_descs)

    def title_repl(match):
        t = match.group(1)
        try:
            ar_t = next(title_iter).replace("'", "\\'")
        except StopIteration:
            ar_t = t
        return f"title: '{t}', arTitle: '{ar_t}'"

    def desc_repl(match):
        d = match.group(1)
        try:
            ar_d = next(desc_iter).replace("'", "\\'")
        except StopIteration:
            ar_d = d
        return f"description: '{d}', arDescription: '{ar_d}'"

    content = re.sub(r"title:\s*'(.*?)'", title_repl, content)
    content = re.sub(r"description:\s*'(.*?)'", desc_repl, content)
    
    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)

    print("Success: translations added to jobs300.js")

if __name__ == "__main__":
    main()
